net-misc/asterisk-base/asterisk-base-2.ebuild

```diff
@@ -0,0 +1,40 @@
+# Copyright 2023 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+inherit systemd tmpfiles
+
+DESCRIPTION="Gentoo Asterisk init scripts"
+HOMEPAGE="https://www.gentoo.org/wiki/No_homepage"
+# Need to set S due to PMS saying we need it existing, but no SRC_URI
+S="${WORKDIR}"
+
+LICENSE="GPL-2"
+SLOT="0"
+KEYWORDS="amd64 ~arm ~arm64 ~ppc ~ppc64 x86"
+
+RDEPEND="
+	net-misc/asterisk
```

Comment:
There appears to be an unbreakable cycle here with asterisk and asterisk-base:
`net-misc/asterisk-base` (RDEPEND) -> `net-misc/asterisk` (RDEPEND) -> `net-misc/asterisk-base` (RDEPEND) -> ...

I think it needs to be a PDEPEND in asterisk-base, or just drop the dep from asterisk-base entirely.

(You could make it a PDEPEND in asterisk, but I think that's a bit less useful. It's fine to have asterisk merged a while after asterisk-base, but you always want asterisk-base ready before asterisk, I'd say?)

Reply:
I only noticed this because of an experimental pkgcheck patch Arthur is working on, but manually verifying it, it makes sense.

What I don't get is why Portage doesn't balk at it:
```
$ emerge -pv1 asterisk-base --tree

These are the packages that would be merged, in reverse order:

Calculating dependencies... done!
Dependency resolution took 7.44 s.

[nomerge       ] net-misc/asterisk-base-1::gentoo
[ebuild  N     ]  net-misc/asterisk-18.18.1:0/18::gentoo  USE="alsa bluetooth caps iconv ssl systemd vorbis -blocks -calendar -cluster -codec2 -curl -dahdi -debug -deprecated -doc -freetds -gtalk -http -ilbc -ldap -lua -mysql -newt -odbc -oss -pjproject -portaudio -postgres -radius (-selinux) -snmp -span -speex -srtp -static -statsd -syslog -unbound -xmpp" LUA_SINGLE_TARGET="lua5-1 -lua5-3 -lua5-4" VOICEMAIL_STORAGE="-imap -odbc" 0 KiB
[ebuild  N     ]   net-misc/asterisk-base-1::gentoo  0 KiB
[ebuild  N     ]   acct-user/asterisk-0-r2::gentoo  0 KiB
[ebuild  N     ]   net-misc/asterisk-moh-opsound-2.03-r1::gentoo  USE="gsm -alaw -g722 -g729 -siren7 -siren14 -sln16 -ulaw -wav" 0 KiB
[ebuild  N     ]   net-misc/asterisk-extra-sounds-1.5.2::gentoo  USE="gsm -alaw -g722 -g729 -siren7 -siren14 -sln16 -ulaw -wav" L10N="en-GB -fr" 3,083 KiB
[ebuild  N     ]   net-misc/asterisk-core-sounds-1.6.1-r1::gentoo  USE="gsm -alaw -g722 -g729 -siren7 -siren14 -sln16 -ulaw -wav" L10N="en-GB -en-AU -es -fr -it -ja -ru -sv" 2,339 KiB
[ebuild  N     ]   acct-group/asterisk-0-r2::gentoo  0 KiB
```

... which may be something I need to look at in its own right.

Reply:
RDEPEND means you don't have to install PRIOR to merging.

At a point there was given a PDEPEND which I questioned precisely because I felt that:

DEPEND means *install prior* to merging this package; and
RDEPEND means this package can *install without*, but *needs* to RUN.

Reply:
Yes, in theory, but [Portage doesn't implement that RDEPEND optimisation](https://bugs.gentoo.org/906809), and the behaviour of RDEPEND wrt circular dependencies is unspecified.

Per https://dev.gentoo.org/~ulm/pms/head/pms.html#section-8.1
> Runtime dependencies (RDEPEND). These must be installed and usable before the results of an ebuild merging are treated as usable.
> Post dependencies (PDEPEND). These must be installed at some point before the package manager finishes the batch of installs.

In this case, I think asterisk-base should either not depend on asterisk at all, or should PDEPEND on it.

Reply:
technically I think that asterisk can stand without -base (nothing stops you from starting it by hand), but -base cannot stand without asterisk.

Re-read that RDEPEND *before the results of an ebuild merging are treated as usable*, which means that RDEPENDS need not be merged prior to the package but at some poit before the package manager finishes the batch of installs, thus effectively RDEPEND is the same as (or at the very least very similar to) PDEPEND.  Again - why have two different options?

Reply:
Sam, if you insist I'm happy to update one or both to PDEPEND, reading the above asterisk is definitely a RDEPEND of asterisk-base (asterisk-base is not USABLE without asterisk), but asterisk itself can PDEPEND on asterisk-base since *technically* no single binary internally to asterisk really depends on asterisk-base, however, you need asterisk-base or init scripts and related (ie, just needs to be merged in the same batch).

Based on that, I argue that we merge this, and if it's a problem, we update asterisk on the next bump to PDEPEND on asterisk-base rather than asterisk-base to PDEPEND onto asterisk.

Reply:
In general I dislike "text files" depending on something, even though in practice this package is useless without asterisk. 

++ for making asterisk PDEPEND on these files. We can revbump and fix asterisks while merging this, if you don't wish to do it here.

Reply:
@juippis, I was hoping to leave this as is for now, portage doesn't choke on this (which @thesamesam is surprised about, but I'm not - because above, probably because I'm not familiar with portage internals).  I get (and I've got bugs open about similar) why this can be a problem, but IMHO the problem is loops in DEPEND, not RDEPEND.

For asterisk side, sure I can tac on an additional commit to this PR, do we need to rev-bump already installed asterisk though?  If you already have asterisk installed this shouldn't be a problem, so for those versions that currently RDEPEND on asterisk-base, "fix" in-place to PDEPEND on asterisk-base to avoid the pointless remerge?

Reply:
Hmmm, yeah I guess it'll work without revbumping. Indeed not every asterisk even depend on -base at the moment, but as I see those handle the init files in their respective ebuilds. There shouldn't be danger of -base being depcleaned.

Reply:
Correct.  Gradual descent.

Updated.  Don't necessarily agree with the change, but I get the motivation (and the implied portage restriction/risk).  I've left the sounds files as RDEPEND since as per the definition asterisk really do need those to operate *normally*, PDEPEND here is OK because these scripts just makes it easier to operate asterisk and isn't strictly speaking required for operation even though it will be tough, and thus IMHO asterisk shouldn't be considered operational until they've been merged (ie, RDEPEND).